File: 0242-Valid-Anagram/0242.java
import java.util.HashMap;
import java.util.Map;

class Solution {
    public boolean isAnagram(String s, String t) {
        if (s.length() != t.length())
            return false;

        Map<Character, Integer> mapS = new HashMap<>();
        Map<Character, Integer> mapT = new HashMap<>();

        for (int i = 0; i < s.length(); i++) {
            mapS.put(s.charAt(i), mapS.getOrDefault(s.charAt(i), 0) + 1);
            mapT.put(t.charAt(i), mapT.getOrDefault(t.charAt(i), 0) + 1);
        }

        for (var pair : mapS.entrySet()) {
            Character c = pair.getKey();
            Integer times = pair.getValue();
            if (!mapT.containsKey(c) || !mapT.get(c).equals(times))
                return false;
        }

        return true;
    }

    public boolean isAnagram1(String s, String t) {
        if (s.length() != t.length())
            return false;

        // the default value in int array is 0
        int[] char1 = new int[26];
        int[] char2 = new int[26];

        for (int i = 0; i < s.length(); i++) {
            char1[s.charAt(i) - 'a']++;
            char2[t.charAt(i) - 'a']++;
        }

        for (int i = 1; i < 26; i++) {
            if (char1[i] != char2[i])
                return false;
        }

        return true;
    }

}

File: 0242-Valid-Anagram/0242.py
class Solution:
    def isAnagram(self, s: str, t: str) -> bool:
        if len(s) != len(t):
            return False
        # key, value = char, times
        map = {}

        for char in s:
            if char in map:
                map[char] += 1
            else:
                map[char] = 1

        for char, times in map.items():
            if t.count(char) != times:
                return False
        
        return True
    
    def isAnagram1(self, s: str, t: str) -> bool:
        return sorted(s) == sorted(t)